### PR TITLE
[ST] Adding STM32 platform and lighting app example with STM32WB5MM-DK board

### PR DESCRIPTION
**Pull request by STYoannZamaron**
> Adding STM32 platform and lighting app example with STM32WB5MM-DK board.
>
> * Add ST as a third party.
> 
> * Add stm32 platform. 
>
> * Add lighting app example with STM32WB5MM-DK over thread.
>
> * Add CI for stm32 example.
> 
> Tests
> lighting app has been tested with multiples controllers such as chip-tool, google home, apple home, smartthings station.
> lighting app is under pre-certification in ST laboratory.

### CODE REVIEW

File: examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+#ifndef CHIP_DEVICE_CONFIG_USE_TEST_SETUP_PIN_CODE
+#define CHIP_DEVICE_CONFIG_USE_TEST_SETUP_PIN_CODE 20202021
+#endif
+
+#ifndef CHIP_DEVICE_CONFIG_USE_TEST_SETUP_DISCRIMINATOR
+#define CHIP_DEVICE_CONFIG_USE_TEST_SETUP_DISCRIMINATOR 0xF00
+#endif
+
+// Use a default pairing code if one hasn't been provisioned in flash.
+#define CHIP_DEVICE_CONFIG_USE_TEST_PAIRING_CODE "CHIPUS"
```
Comment (bzbarsky-apple) [inline examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h:37]: This define seems to be unused.  Why is it being defined?

File: examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+/**
+ * CHIP_DEVICE_CONFIG_ENABLE_TEST_DEVICE_IDENTITY
+ *
+ * Enables the use of a hard-coded default Chip device id and credentials if no device id
+ * is found in Chip NV storage.
+ *
+ * This option is for testing only and should be disabled in production releases.
+ */
+#define CHIP_DEVICE_CONFIG_ENABLE_TEST_DEVICE_IDENTITY 34
```
Comment (bzbarsky-apple) [inline examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h:95]: This also looks unused.  Why is it here?

File: examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+// For convenience, enable Chip Security Test Mode and disable the requirement for
+// authentication in various protocols.
+//
+//    WARNING: These options make it possible to circumvent basic Chip security functionality,
+//    including message encryption. Because of this they MUST NEVER BE ENABLED IN PRODUCTION BUILDS.
+//
+#define CHIP_CONFIG_SECURITY_TEST_MODE 0
```
Comment (bzbarsky-apple) [inline examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h:103]: This is already defined above....

File: examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+#define CHIP_CONFIG_REQUIRE_AUTH 1
```
Comment (bzbarsky-apple) [inline examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h:104]: This is unused.

File: docs/guides/stm32_getting_started_guide.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+## Matter Example Applications
+
+Sample Matter applications are provided for the stm32 platform. They can be used
+to speed up development. You can find them in the samples with `/stm32`
+subfolder.
+
+## Building the Example Application
+
+-   [Setup Matter Environment](./BUILDING.md)
```
Comment (bzbarsky-apple) [inline docs/guides/stm32_getting_started_guide.md:31]:
```suggestion
-   [Set Up Matter Environment](./BUILDING.md)
```

File: docs/guides/stm32_getting_started_guide.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+-   Setup STLINK tools
```
Comment (bzbarsky-apple) [inline docs/guides/stm32_getting_started_guide.md:33]:
```suggestion
-   Set up STLINK tools
```
Or whatever capitalization you want here; either "Up" and "Tools", or "up" and "tools"

File: src/platform/stm32/DiagnosticDataProviderImpl.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+DiagnosticDataProviderImpl & DiagnosticDataProviderImpl::GetDefaultInstance()
+{
+    static DiagnosticDataProviderImpl sInstance;
+    return sInstance;
+}
+
+CHIP_ERROR DiagnosticDataProviderImpl::GetCurrentHeapFree(uint64_t & currentHeapFree)
+{
+
+    return CHIP_NO_ERROR;
```
Comment (bzbarsky-apple) [inline src/platform/stm32/DiagnosticDataProviderImpl.cpp:42]: This is broken.  Please do not return CHIP_NO_ERROR while not setting the outparams.  That applies to a number of functions in this platform implementation.

File: src/platform/stm32/PlatformManagerImpl.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+CHIP_ERROR PlatformManagerImpl::_Shutdown()
+{
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+}
+
+CHIP_ERROR PlatformManagerImpl::_GetCurrentHeapFree(uint64_t & currentHeapFree)
+{
+
+	return CHIP_NO_ERROR;
```
Comment (bzbarsky-apple) [inline src/platform/stm32/PlatformManagerImpl.cpp:62]: More broken functions here and following.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/lighting-app/stm32/include/STM32WB5/CHIPProjectConfig.h:37]: And I know you copy/pasted this, but see https://github.com/project-chip/connectedhomeip/pull/28906

File: .github/.wordlist.txt [hunk 9]
```diff
@@ -1276,6 +1276,10 @@ StatusCode
 stderr
 stdout
 sterm
+STMicroelectronics
+stm
```
[the remark this answers is not in the document]
Comment (STYoannZamaron) [reply .github/.wordlist.txt:1280]: ok thanks

File: src/platform/stm32/DiagnosticDataProviderImpl.cpp (hunk 7 above)
Comment (STYoannZamaron) [reply src/platform/stm32/DiagnosticDataProviderImpl.cpp]: thanks, I fully agree with that.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/platform/stm32/PlatformManagerImpl.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+void PlatformManagerImpl::_RunEventLoop(void)
+{
+    Internal::GenericPlatformManagerImpl_FreeRTOS<PlatformManagerImpl>::_RunEventLoop();
+}
+
+CHIP_ERROR PlatformManagerImpl::_Shutdown()
+{
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
```
[the remark this answers is not in the document]
Comment (STYoannZamaron) [reply src/platform/stm32/PlatformManagerImpl.cpp:54]: for now, we didn't implement the shutdown at application layer.

[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/platform/stm32/PlatformManagerImpl.cpp:54]: This should probably still call shutdown on the super-class and then return error if it wants to....

File: scripts/build/builders/stm32.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+class stm32Builder(GnBuilder):
+
+    def __init__(self,
+                 root,
+                 runner,
+                 app: stm32App = stm32App.LIGHT,
+                 board: stm32Board = stm32Board.STM32WB55XX,
+                 chip_build_libshell: bool = False,
+                 chip_logging: bool = True,
+                 enable_factory: bool = False,
```
Comment (andy31415) [inline scripts/build/builders/stm32.py:65]: please remove unused arguments - I see no use for logging/factory/shell/rotating_device_id/requestor/hook/rpcs

**Review by andy31415 — APPROVED** [review]

File: src/platform/stm32/PlatformManagerImpl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (STYoannZamaron) [reply src/platform/stm32/PlatformManagerImpl.cpp]: done

**Pull request merged**